README.md
# `eos-common`

[![Build Status](https://travis-ci.org/EOS-Nation/eos-common.svg?branch=master)](https://travis-ci.org/EOS-Nation/eos-common)
[![npm version](https://badge.fury.io/js/eos-common.svg)](https://badge.fury.io/js/eos-common)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/EOS-Nation/eos-common/master/LICENSE)

> EOSIO Smart Contract common library used for Typescript

Implements most commonly used EOSIO C++ Classes into Typescript:

-   [asset](https://github.com/EOSIO/eosio.cdt/blob/master/libraries/eosiolib/asset.hpp)
-   [symbol](https://github.com/EOSIO/eosio.cdt/blob/master/libraries/eosiolib/symbol.hpp)
-   [symbol_code](https://github.com/EOSIO/eosio.cdt/blob/master/libraries/eosiolib/symbol.hpp)
-   [name](https://github.com/EOSIO/eosio.cdt/blob/master/libraries/eosiolib/name.hpp)
-   [extended_asset](https://github.com/EOSIO/eosio.cdt/blob/master/libraries/eosiolib/asset.hpp)
-   [extended_symbol](https://github.com/EOSIO/eosio.cdt/blob/master/libraries/eosiolib/symbol.hpp)

## Installation

Using Yarn:

```bash
yarn add eos-common
```

or using NPM:

```bash
npm install --save eos-common
```

## Quick Start

```ts
import { asset, symbol } from "eos-common"

const quantity = asset("1.0000 EOS");
// or
const quantity = asset(10000, symbol("EOS", 4));
quantity.to_string() //=> "1.0000 EOS";
quantity.symbol.code().to_string() //=> "EOS"
quantity.symbol.precision() //=> 4
```

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [Asset](#asset)
    -   [Parameters](#parameters)
    -   [Examples](#examples)
    -   [amount](#amount)
    -   [symbol](#symbol)
    -   [typeof](#typeof)
    -   [is_amount_within_range](#is_amount_within_range)
    -   [is_valid](#is_valid)
    -   [set_amount](#set_amount)
        -   [Parameters](#parameters-1)
    -   [minus](#minus)
        -   [Parameters](#parameters-2)
    -   [plus](#plus)
        -   [Parameters](#parameters-3)
    -   [times](#times)
        -   [Parameters](#parameters-4)
    -   [div](#div)
        -   [Parameters](#parameters-5)
    -   [toString](#tostring)
    -   [isInstance](#isinstance)
        -   [Parameters](#parameters-6)
    -   [plus](#plus-1)
        -   [Parameters](#parameters-7)
    -   [minus](#minus-1)
        -   [Parameters](#parameters-8)
    -   [times](#times-1)
        -   [Parameters](#parameters-9)
    -   [div](#div-1)
        -   [Parameters](#parameters-10)
    -   [isEqual](#isequal)
        -   [Parameters](#parameters-11)
    -   [isNotEqual](#isnotequal)
        -   [Parameters](#parameters-12)
    -   [isLessThan](#islessthan)
        -   [Parameters](#parameters-13)
    -   [isLessThanOrEqual](#islessthanorequal)
        -   [Parameters](#parameters-14)
    -   [isGreaterThan](#isgreaterthan)
        -   [Parameters](#parameters-15)
    -   [isGreaterThanOrEqual](#isgreaterthanorequal)
        -   [Parameters](#parameters-16)
    -   [max_amount](#max_amount)
-   [check](#check)
    -   [Parameters](#parameters-17)
    -   [Examples](#examples-1)
-   [write_decimal](#write_decimal)
    -   [Parameters](#parameters-18)
-   [ExtendedAsset](#extendedasset)
    -   [Parameters](#parameters-19)
    -   [quantity](#quantity)
    -   [contract](#contract)
    -   [typeof](#typeof-1)
    -   [get_extended_symbol](#get_extended_symbol)
    -   [toString](#tostring-1)
    -   [times](#times-2)
        -   [Parameters](#parameters-20)
    -   [div](#div-2)
        -   [Parameters](#parameters-21)
    -   [minus](#minus-2)
        -   [Parameters](#parameters-22)
    -   [plus](#plus-2)
        -   [Parameters](#parameters-23)
    -   [isLessThan](#islessthan-1)
        -   [Parameters](#parameters-24)
    -   [isEqual](#isequal-1)
        -   [Parameters](#parameters-25)
    -   [isNotEqual](#isnotequal-1)
        -   [Parameters](#parameters-26)
    -   [isLessThanOrEqual](#islessthanorequal-1)
        -   [Parameters](#parameters-27)
    -   [isGreaterThanOrEqual](#isgreaterthanorequal-1)
        -   [Parameters](#parameters-28)
    -   [isInstance](#isinstance-1)
        -   [Parameters](#parameters-29)
-   [ExtendedSymbol](#extendedsymbol)
    -   [Parameters](#parameters-30)
    -   [typeof](#typeof-2)
    -   [get_symbol](#get_symbol)
    -   [get_contract](#get_contract)
    -   [toString](#tostring-2)
        -   [Parameters](#parameters-31)
    -   [isInstance](#isinstance-2)
        -   [Parameters](#parameters-32)
    -   [isEqual](#isequal-2)
        -   [Parameters](#parameters-33)
    -   [isNotEqual](#isnotequal-2)
        -   [Parameters](#parameters-34)
    -   [isLessThan](#islessthan-2)
        -   [Parameters](#parameters-35)
-   [extended_symbol](#extended_symbol)
    -   [Parameters](#parameters-36)
    -   [Examples](#examples-2)
-   [Name](#name)
    -   [Parameters](#parameters-37)
    -   [typeof](#typeof-3)
    -   [length](#length)
    -   [suffix](#suffix)
    -   [raw](#raw)
    -   [bool](#bool)
    -   [toString](#tostring-3)
    -   [isInstance](#isinstance-3)
        -   [Parameters](#parameters-38)
    -   [char_to_value](#char_to_value)
        -   [Parameters](#parameters-39)
    -   [isEqual](#isequal-3)
        -   [Parameters](#parameters-40)
    -   [isNotEqual](#isnotequal-3)
        -   [Parameters](#parameters-41)
    -   [isLessThan](#islessthan-3)
        -   [Parameters](#parameters-42)
-   [SymbolCode](#symbolcode)
    -   [Parameters](#parameters-43)
    -   [typeof](#typeof-4)
    -   [toString](#tostring-4)
    -   [isEqual](#isequal-4)
        -   [Parameters](#parameters-44)
    -   [isNotEqual](#isnotequal-4)
        -   [Parameters](#parameters-45)
    -   [isLessThan](#islessthan-4)
        -   [Parameters](#parameters-46)
    -   [isInstance](#isinstance-4)
        -   [Parameters](#parameters-47)
-   [Symbol](#symbol-1)
    -   [Parameters](#parameters-48)
    -   [Examples](#examples-3)
-   [typeof](#typeof-5)
-   [is_valid](#is_valid-1)
-   [precision](#precision)
-   [code](#code)
-   [raw](#raw-1)
-   [bool](#bool-1)
-   [toString](#tostring-5)
    -   [Parameters](#parameters-49)
-   [isInstance](#isinstance-5)
    -   [Parameters](#parameters-50)
-   [isEqual](#isequal-5)
    -   [Parameters](#parameters-51)
-   [isNotEqual](#isnotequal-5)
    -   [Parameters](#parameters-52)
-   [isLessThan](#islessthan-5)
    -   [Parameters](#parameters-53)
-   [voteWeightToday](#voteweighttoday)
-   [stake2vote](#stake2vote)
    -   [Parameters](#parameters-54)
-   [vote2stake](#vote2stake)
    -   [Parameters](#parameters-55)
-   [calculate_producer_per_vote_pay](#calculate_producer_per_vote_pay)
    -   [Parameters](#parameters-56)

### Asset

Asset

#### Parameters

-   `amount` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** The amount of the asset
-   `sym` **[Symbol](#symbol)** The name of the symbol

#### Examples

```javascript
const quantity = new Asset(10000, new Symbol("EOS", 4));
quantity.toString() //=> "1.0000 EOS";
quantity.symbol.code() //=> "EOS"
quantity.symbol.precision //=> 4
```

Returns **[Asset](#asset)** Asset

#### amount

{int64_t} The amount of the asset

#### symbol

{symbol} The symbol name of the asset

#### typeof

The typeof operator returns a string indicating the type of the unevaluated operand.

#### is_amount_within_range

Check if the amount doesn't exceed the max amount

Returns **any** true - if the amount doesn't exceed the max amount

Returns **any** false - otherwise

#### is_valid

Check if the asset is valid. %A valid asset has its amount &lt;= max_amount and its symbol name valid

Returns **any** true - if the asset is valid

Returns **any** false - otherwise

#### set_amount

Set the amount of the asset

##### Parameters

-   `amount`  
-   `a`  New amount for the asset

#### minus

Subtraction assignment operator

##### Parameters

-   `a`  Another asset to subtract this asset with

Returns **any** asset& - Reference to this asset

#### plus

Addition Assignment  operator

##### Parameters

-   `a`  Another asset to subtract this asset with

Returns **any** asset& - Reference to this asset

#### times

Multiplication assignment operator, with a number

##### Parameters

-   `a`  The multiplier for the asset's amount

Returns **any** asset - Reference to this asset

#### div

##### Parameters

-   `a`  The divisor for the asset's amount

Returns **any** asset - Reference to this asset

#### toString

The toString() method returns the string representation of the object.

#### isInstance

The isinstance() function returns True if the specified object is of the specified type, otherwise False.

##### Parameters

-   `obj`  

#### plus

Addition operator

##### Parameters

-   `a`  The first asset to be added
-   `b`  The second asset to be added

Returns **any** asset - New asset as the result of addition

#### minus

Subtraction operator

##### Parameters

-   `a`  The asset to be subtracted
-   `b`  The asset used to subtract

Returns **any** asset - New asset as the result of subtraction of a with b

#### times

Multiplication operator, with a number proceeding

##### Parameters

-   `a`  The asset to be multiplied
-   `b`  The multiplier for the asset's amount

Returns **any** asset - New asset as the result of multiplication

#### div

Division operator, with a number proceeding

##### Parameters

-   `a`  The asset to be divided
-   `b`  The divisor for the asset's amount

Returns **any** asset - New asset as the result of division

#### isEqual

Equality operator

##### Parameters

-   `a`  The first asset to be compared
-   `b`  The second asset to be compared

Returns **any** true - if both asset has the same amount

Returns **any** false - otherwise

#### isNotEqual

Inequality operator

##### Parameters

-   `a`  The first asset to be compared
-   `b`  The second asset to be compared

Returns **any** true - if both asset doesn't have the same amount

Returns **any** false - otherwise

#### isLessThan

Less than operator

##### Parameters

-   `a`  The first asset to be compared
-   `b`  The second asset to be compared

Returns **any** true - if the first asset's amount is less than the second asset amount

Returns **any** false - otherwise

#### isLessThanOrEqual

Less or equal to operator

##### Parameters

-   `a`  The first asset to be compared
-   `b`  The second asset to be compared

Returns **any** true - if the first asset's amount is less or equal to the second asset amount

Returns **any** false - otherwise

#### isGreaterThan

Greater than operator

##### Parameters

-   `a`  The first asset to be compared
-   `b`  The second asset to be compared

Returns **any** true - if the first asset's amount is greater than the second asset amount

Returns **any** false - otherwise

#### isGreaterThanOrEqual

Greater or equal to operator

##### Parameters

-   `a`  The first asset to be compared
-   `b`  The second asset to be compared

Returns **any** true - if the first asset's amount is greater or equal to the second asset amount

Returns **any** false - otherwise

#### max_amount

{constexpr int64_t} Maximum amount possible for this asset. It's capped to 2^62 - 1

### check

Assert if the predicate fails and use the supplied message.

#### Parameters

-   `pred` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Pre-condition
-   `msg` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Error Message

#### Examples

```javascript
check(a == b, "a does not equal b");
```

Returns **void** 

### write_decimal

Writes a number as a string

#### Parameters

-   `number`  The number to print before shifting the decimal point to the left by num_decimal_places.
-   `num_decimal_places`  The number of decimal places to shift the decimal point.
-   `negative`  Whether to print a minus sign in the front.

### ExtendedAsset

#### Parameters

-   `obj1`  
-   `obj2`  

#### quantity

The asset

#### contract

The owner of the asset

#### typeof

The typeof operator returns a string indicating the type of the unevaluated operand.

#### get_extended_symbol

Get the extended symbol of the asset

Returns **any** extended_symbol - The extended symbol of the asset

#### toString

The toString() method returns the string representation of the object.

#### times

Multiplication assignment operator

##### Parameters

-   `a`  

#### div

Division operator

##### Parameters

-   `a`  

#### minus

Subtraction operator

##### Parameters

-   `a`  

#### plus

Addition operator

##### Parameters

-   `a`  

#### isLessThan

Less than operator

##### Parameters

-   `a`  

#### isEqual

Comparison operator

##### Parameters

-   `a`  

#### isNotEqual

Comparison operator

##### Parameters

-   `a`  

#### isLessThanOrEqual

Comparison operator

##### Parameters

-   `a`  

#### isGreaterThanOrEqual

Comparison operator

##### Parameters

-   `a`  

#### isInstance

The isinstance() function returns True if the specified object is of the specified type, otherwise False.

##### Parameters

-   `obj`  

### ExtendedSymbol

#### Parameters

-   `sym`  The symbol
-   `contract`  
-   `con`  The name of the contract

#### typeof

The typeof operator returns a string indicating the type of the unevaluated operand.

#### get_symbol

Returns the symbol in the extended_contract

Returns **any** symbol

#### get_contract

Returns the name of the contract in the extended_symbol

Returns **any** name

#### toString

The toString() method returns the string representation of the object.

##### Parameters

-   `show_precision`   (optional, default `true`)

#### isInstance

The isinstance() function returns True if the specified object is of the specified type, otherwise False.

##### Parameters

-   `obj`  

#### isEqual

Equivalency operator. Returns true if a == b (are the same)

##### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if both provided name are the same

#### isNotEqual

Inverted equivalency operator. Returns true if a != b (are different)

##### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if both provided name are not the same

#### isLessThan

Less than operator. Returns true if a &lt; b.

##### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if name `a` is less than `b`

### extended_symbol

Extended Symbol

#### Parameters

-   `sym`  
-   `contract`  

#### Examples

```javascript
extended_symbol( symbol("4,EOS"), name("eosio.token") )
```

### Name

#### Parameters

-   `str`  

#### typeof

The typeof operator returns a string indicating the type of the unevaluated operand.

#### length

Returns the length of the %name

#### suffix

Returns the suffix of the %name

#### raw

Returns uint64_t repreresentation of the name

#### bool

Explicit cast to bool of the name

Returns **any** Returns true if the name is set to the default value of 0 else true.

#### toString

The toString() method returns the string representation of the object.

#### isInstance

The isinstance() function returns True if the specified object is of the specified type, otherwise False.

##### Parameters

-   `obj`  

#### char_to_value

Converts a %name Base32 symbol into its corresponding value

##### Parameters

-   `c`  Character to be converted

Returns **any** constexpr char - Converted value

#### isEqual

Equivalency operator. Returns true if a == b (are the same)

##### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if both provided name are the same

#### isNotEqual

Inverted equivalency operator. Returns true if a != b (are different)

##### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if both provided name are not the same

#### isLessThan

Less than operator. Returns true if a &lt; b.

##### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if name `a` is less than `b`

### SymbolCode

#### Parameters

-   `str`  

#### typeof

The typeof operator returns a string indicating the type of the unevaluated operand.

#### toString

The toString() method returns the string representation of the object.

#### isEqual

Equivalency operator. Returns true if a == b (are the same)

##### Parameters

-   `comparison`  

Returns **any** boolean - true if both provided symbol_codes are the same

#### isNotEqual

Inverted equivalency operator. Returns true if a != b (are different)

##### Parameters

-   `comparison`  

Returns **any** boolean - true if both provided symbol_codes are not the same

#### isLessThan

Less than operator. Returns true if a &lt; b.

##### Parameters

-   `comparison`  

Returns **any** boolean - true if symbol_code `a` is less than `b`

#### isInstance

The isinstance() function returns True if the specified object is of the specified type, otherwise False.

##### Parameters

-   `obj`  

### Symbol

Symbol

#### Parameters

-   `code` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Symbol Code
-   `precision` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Precision

#### Examples

```javascript
const sym = new Symbol("EOS", 4);
sym.code() //=> "EOS"
sym.precision //=> 4
```

Returns **[Symbol](#symbol)** Symbol

### typeof

The typeof operator returns a string indicating the type of the unevaluated operand.

### is_valid

Is this symbol valid

### precision

This symbol's precision

### code

Returns representation of symbol name

### raw

Returns uint64_t repreresentation of the symbol

### bool

Explicit cast to bool of the symbol

Returns **any** Returns true if the symbol is set to the default value of 0 else true.

### toString

The toString() method returns the string representation of the object.

#### Parameters

-   `show_precision`   (optional, default `true`)

### isInstance

The isinstance() function returns True if the specified object is of the specified type, otherwise False.

#### Parameters

-   `obj`  

### isEqual

Equivalency operator. Returns true if a == b (are the same)

#### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if both provided symbol_codes are the same

### isNotEqual

Inverted equivalency operator. Returns true if a != b (are different)

#### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if both provided symbol_codes are not the same

### isLessThan

Less than operator. Returns true if a &lt; b.

#### Parameters

-   `a`  
-   `b`  

Returns **any** boolean - true if symbol_code `a` is less than `b`

### voteWeightToday

voteWeightToday computes the stake2vote weight for EOS, in order to compute the decaying value.

### stake2vote

Convert EOS stake into decaying value

#### Parameters

-   `staked`  
-   `vote` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** vote

### vote2stake

Convert vote decay value into EOS stake

#### Parameters

-   `vote`  
-   `staked` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** staked

### calculate_producer_per_vote_pay

Calculate producer vpay

#### Parameters

-   `total_votes`  
-   `pervote_bucket`  
-   `total_producer_vote_weight`  

Returns **bigint** producer pay as int64t


CHANGES.md
## 2020-05-01

- add `extended_symbol.raw()`

## 2020-04-06

- support adding `asset` as param to `asset`

## 2020-03-07

- rename `Symbol` => `Sym` (conflict with native Symbol class)
- added `name` Class
- added `extended_symbol` Class

## 2020-03-06

- Add operators to `asset`
  - max_amount
  - plus
  - minus
  - times
  - div
  - isEqual
  - isNotEqual
  - isLessThan
  - isLessThanOrEqual
  - isGreaterThan
  - isGreaterThanOrEqual

## 2020-02-25
- add `symbol_code` unit tests from C++ tests
- add `symbol` method
- add `symbol_code` method
- fix 7 character symcode
- fully refactor `Symbol` to C++ represenation

## 2020-02-24

- `toNumber` => `to_double`
- `toString` => `to_string`
- remove `toDecimal`
- remove `decimal.js` dependency
- added rollup for browser bundling


eosiolib/asset.ts
import { Sym, symbol } from "./symbol";
import { check } from "./check";
import { write_decimal } from "./eosiolib";
import { getSymbol, getAmount, isNull, number_to_bigint, getType } from "./utils";
import bigInt, { BigInteger } from "big-integer";

/**
 * Asset
 *
 * @name Asset
 * @param {number} amount The amount of the asset
 * @param {Symbol} sym  The name of the symbol
 * @returns {Asset} Asset
 * @example
 *
 * const quantity = new Asset(10000, new Symbol("EOS", 4));
 * quantity.toString() //=> "1.0000 EOS";
 * quantity.symbol.code() //=> "EOS"
 * quantity.symbol.precision //=> 4
 */
export class Asset {
    get [Symbol.toStringTag](): string {
        return 'asset';
    }
    /**
     * The typeof operator returns a string indicating the type of the unevaluated operand.
     */
    public get typeof(): string { return 'asset' }

    /**
     * The isinstance() function returns True if the specified object is of the specified type, otherwise False.
     */
    public static isInstance( obj: any ): boolean { return obj instanceof Asset; }

    /**
     * {constexpr int64_t} Maximum amount possible for this asset. It's capped to 2^62 - 1
     */
    public static max_amount = (bigInt(1).shiftLeft(62)).minus(1);

    /**
     * {int64_t} The amount of the asset
     */
    public amount = bigInt(0);

    /**
     * {symbol} The symbol name of the asset
     */
    public symbol: Sym = symbol();

    /**
     * Construct a new asset given the symbol name and the amount
     */
    constructor ( obj1?: string | number | BigInteger | bigint | Asset, obj2?: Sym ) {
        if ( isNull(obj1) && isNull(obj2) ) {
            return;
        }
        if ( typeof obj1 == "string" ) {
            const [amount_str, symbol_str] = obj1.split(" ");
            const precision = (amount_str.split(".")[1] || []).length;
            this.amount = number_to_bigint( Number( amount_str ) * Math.pow(10, precision));
            this.symbol = new Sym( symbol_str, precision );
        } else if ( getType( obj1 ) == "asset") {
            const _sym = getSymbol( obj1 );
            if ( !_sym ) throw new Error("[sym] is required");

            this.amount = getAmount( obj1 );
            this.symbol = _sym;
        } else if ( obj2 ) {
            this.amount = getAmount( obj1 );
            this.symbol = obj2;
        } else {
            throw new Error("[sym] is required");
        }

        check( this.is_amount_within_range(), "magnitude of asset amount must be less than 2^62" );
        check( this.symbol.is_valid(), "invalid symbol name" );
    }

    /**
     * Check if the amount doesn't exceed the max amount
     *
     * @return true - if the amount doesn't exceed the max amount
     * @return false - otherwise
     */
    public is_amount_within_range(): boolean {
        return (Asset.max_amount.multiply(-1)).lesserOrEquals(this.amount) && this.amount.lesserOrEquals(Asset.max_amount);
    }

    /**
     * Check if the asset is valid. %A valid asset has its amount <= max_amount and its symbol name valid
     *
     * @return true - if the asset is valid
     * @return false - otherwise
     */
    public is_valid(): boolean {
        return this.is_amount_within_range() && this.symbol.is_valid();
    }

    /**
     * Set the amount of the asset
     *
     * @param a - New amount for the asset
     */
    public set_amount( amount: BigInteger | number ): void {
        this.amount = typeof amount == "number" ? bigInt(amount) : amount;
        check( this.is_amount_within_range(), "magnitude of asset amount must be less than 2^62" );
    }

    /**
     * Subtraction assignment operator
     *
     * @param a - Another asset to subtract this asset with
     * @return asset& - Reference to this asset
     * @post The amount of this asset is subtracted by the amount of asset a
     */
    public minus( a: Asset | number | bigint | BigInteger ): Asset {
        const amount = getAmount( a );
        const sym = getSymbol( a );
        if ( sym ) check( sym.isEqual( this.symbol ), "attempt to subtract asset with different symbol" );

        this.amount = this.amount.minus( amount );
        check( Asset.max_amount.multiply(-1).lesserOrEquals(this.amount), "subtraction underflow" );
        check( this.amount.lesserOrEquals(Asset.max_amount),  "subtraction overflow" );
        return this;
    }

    /**
     * Addition Assignment  operator
     *
     * @param a - Another asset to subtract this asset with
     * @return asset& - Reference to this asset
     * @post The amount of this asset is added with the amount of asset a
     */
    public plus( a: Asset | number | bigint | BigInteger ): Asset {
        const amount = getAmount( a );
        const sym = getSymbol( a );
        if ( sym ) check( sym.isEqual( this.symbol ), "attempt to add asset with different symbol" );

        this.amount = this.amount.plus( amount );
        check( Asset.max_amount.multiply(-1).lesserOrEquals(this.amount), "addition underflow" );
        check( this.amount.lesserOrEquals( Asset.max_amount ),  "addition overflow" );
        return this;
    }

    /**
     * Addition operator
     *
     * @param a - The first asset to be added
     * @param b - The second asset to be added
     * @return asset - New asset as the result of addition
     */
    public static plus( a: Asset, b: Asset | number | bigint | BigInteger ): Asset {
        const result = new Asset(a.amount, a.symbol);
        result.plus( b );
        return result;
    }

    /**
     * Subtraction operator
     *
     * @param a - The asset to be subtracted
     * @param b - The asset used to subtract
     * @return asset - New asset as the result of subtraction of a with b
     */
    public static minus( a: Asset, b: Asset | number | bigint | BigInteger ): Asset {
        const result = new Asset(a.amount, a.symbol);
        result.minus( b );
        return result;
    }

    /**
     * Multiplication assignment operator, with a number
     *
     * @details Multiplication assignment operator. Multiply the amount of this asset with a number and then assign the value to itself.
     * @param a - The multiplier for the asset's amount
     * @return asset - Reference to this asset
     * @post The amount of this asset is multiplied by a
     */
    public times( a: Asset | number | bigint | BigInteger ): Asset {
        const amount = getAmount( a );
        const sym = getSymbol( a );
        if ( sym ) check( sym.isEqual( this.symbol ), "comparison of assets with different symbols is not allowed" );

        const tmp = this.amount.multiply(amount);
        check( tmp.lesserOrEquals( Asset.max_amount), "multiplication overflow" );
        check( tmp.greaterOrEquals(Asset.max_amount.multiply(-1)), "multiplication underflow" );
        this.amount = tmp;
        return this;
    }

    /**
     * Multiplication operator, with a number proceeding
     *
     * @brief Multiplication operator, with a number proceeding
     * @param a - The asset to be multiplied
     * @param b - The multiplier for the asset's amount
     * @return asset - New asset as the result of multiplication
     */
    public static times( a: Asset, b: Asset | number | bigint | BigInteger ): Asset {
        const result = new Asset(a.amount, a.symbol);
        result.times(b);
        return result;
    }

    /**
     * @brief Division assignment operator, with a number
     *
     * @details Division assignment operator. Divide the amount of this asset with a number and then assign the value to itself.
     * @param a - The divisor for the asset's amount
     * @return asset - Reference to this asset
     * @post The amount of this asset is divided by a
     */
    public div( a: Asset | number | bigint | BigInteger ): Asset {
        const amount = getAmount( a );
        const sym = getSymbol( a );
        if ( sym ) check( sym.isEqual( this.symbol ), "comparison of assets with different symbols is not allowed" );

        check( amount.notEquals(0), "divide by zero" );
        check( !(this.amount.equals(Asset.max_amount.multiply(-1)) && amount.equals(-1)), "signed division overflow" );
        this.amount = this.amount.divide(amount);
        return this;
    }

    /**
     * Division operator, with a number proceeding
     *
     * @param a - The asset to be divided
     * @param b - The divisor for the asset's amount
     * @return asset - New asset as the result of division
     */
    public static div( a: Asset, b: Asset | number | bigint | BigInteger ): Asset {
        const result = new Asset( a.amount, a.symbol );
        result.div(b);
        return result;
    }

    /**
     * Equality operator
     *
     * @param a - The first asset to be compared
     * @param b - The second asset to be compared
     * @return true - if both asset has the same amount
     * @return false - otherwise
     * @pre Both asset must have the same symbol
     */
    public static isEqual( a: Asset, b: Asset ): boolean {
        check( a.symbol.isEqual( b.symbol ), "comparison of assets with different symbols is not allowed" );
        return a.amount.equals( b.amount );
    }

    public isEqual( a: Asset ): boolean {
        check( a.symbol.isEqual( this.symbol ), "comparison of assets with different symbols is not allowed" );
        return a.amount.equals( this.amount );
    }

    /**
     * Inequality operator
     *
     * @param a - The first asset to be compared
     * @param b - The second asset to be compared
     * @return true - if both asset doesn't have the same amount
     * @return false - otherwise
     * @pre Both asset must have the same symbol
     */
    public static isNotEqual( a: Asset, b: Asset ): boolean {
        return !( a.isEqual(b));
    }

    public isNotEqual( a: Asset ): boolean {
        return !( a.isEqual(this) );
    }

    /**
     * Less than operator
     *
     * @param a - The first asset to be compared
     * @param b - The second asset to be compared
     * @return true - if the first asset's amount is less than the second asset amount
     * @return false - otherwise
     * @pre Both asset must have the same symbol
     */
    public static isLessThan( a: Asset, b: Asset ): boolean {
        check( a.symbol.isEqual( b.symbol ), "comparison of assets with different symbols is not allowed" );
        return a.amount.lesser(b.amount);
    }

    public isLessThan( a: Asset ): boolean {
        check( a.symbol.isEqual( this.symbol ), "comparison of assets with different symbols is not allowed" );
        return this.amount.lesser(a.amount);
    }

    /**
     * Less or equal to operator
     *
     * @param a - The first asset to be compared
     * @param b - The second asset to be compared
     * @return true - if the first asset's amount is less or equal to the second asset amount
     * @return false - otherwise
     * @pre Both asset must have the same symbol
     */
    public static isLessThanOrEqual( a: Asset, b: Asset ): boolean {
        check( a.symbol.isEqual( b.symbol ), "comparison of assets with different symbols is not allowed" );
        return a.amount.lesserOrEquals(b.amount);
    }

    public isLessThanOrEqual( a: Asset ): boolean {
        check( a.symbol.isEqual( this.symbol ), "comparison of assets with different symbols is not allowed" );
        return this.amount.lesserOrEquals(a.amount);
    }

    /**
     * Greater than operator
     *
     * @param a - The first asset to be compared
     * @param b - The second asset to be compared
     * @return true - if the first asset's amount is greater than the second asset amount
     * @return false - otherwise
     * @pre Both asset must have the same symbol
     */
    public static isGreaterThan( a: Asset, b: Asset ): boolean {
        check( a.symbol.isEqual( b.symbol ), "comparison of assets with different symbols is not allowed" );
        return a.amount.greater( b.amount );
    }

    public isGreaterThan( a: Asset ): boolean {
        check( a.symbol.isEqual( this.symbol ), "comparison of assets with different symbols is not allowed" );
        return this.amount.greater( a.amount );
    }

    /**
     * Greater or equal to operator
     *
     * @param a - The first asset to be compared
     * @param b - The second asset to be compared
     * @return true - if the first asset's amount is greater or equal to the second asset amount
     * @return false - otherwise
     * @pre Both asset must have the same symbol
     */
    public static isGreaterThanOrEqual( a: Asset, b: Asset ): boolean {
        check( a.symbol.isEqual( b.symbol ), "comparison of assets with different symbols is not allowed" );
        return a.amount.greaterOrEquals( b.amount );
    }

    public isGreaterThanOrEqual( a: Asset ): boolean {
        check( a.symbol.isEqual( this.symbol ), "comparison of assets with different symbols is not allowed" );
        return this.amount.greaterOrEquals( a.amount );
    }

    /**
     * The toString() method returns the string representation of the object.
     */
    public toString(): string {
        return this.to_string();
    }

    public to_string(): string {
        const amount = write_decimal(this.amount, this.symbol.precision(), true);
        const symcode = this.symbol.code().to_string();
        return `${amount} ${symcode}`;
    }

    // /**
    //  * %Print the asset
    //  *
    //  * @brief %Print the asset
    //  */
    // public print(): void {
    //     process.stdout.write(this.to_string());
    // }
}

export const asset: {
    /**
     * String
     *
     * @example
     *
     * const quantity = asset( "1.0000 EOS" )
     * asset( quantity )
     */
    ( asset?: Asset | string ): Asset;

    /**
     * Amount & Sym
     *
     * @example
     *
     * asset( 10000, symbol("4,EOS") )
     */
    ( amount?: number | bigint | BigInteger, sym?: Sym ): Asset;
} = ( obj1?: any, obj2?: any ) => {
    return new Asset( obj1, obj2 );
}
